### PR TITLE
Resolve #3913: Missing esp32 status() implementation

### LINKED ISSUE #3913: ESP32: isconnected() and status() methods do not work
Hi,

i have pull the repository a couple of days ago and now i'm on the commit d8dc918d (py/compile: Handle return/break/continue correctly in async with.) of 23/06/2018.
Now when the ESP32 disconnect for some reasons in STA_IF mode, isconnected() method always returns True and the ESP32 log shows as follows:

`I (5476584) wifi: STA_DISCONNECTED, reason:201`

If i call the status() method he always returns None so i'm not able to determinate the status of the network.STA_IF.

Thanks in advance for the help.

### PR DESCRIPTION
**Pull request by mitchins**
Resubmit with cleaner rebase and diff'ing. (GitHub
 closed the old one automatically)

Implements the default method (no args) which provides the link status for the STA_IF on ESP32 devices.

Implementation is inspired and consistent with ESP8266, success codes are named similarly, as are most error names. Error codes and names for wifi disconnect are otherwise taken (returned directly) from what the ESP-IDF gives us but are defined in the const dictionary where appropriate.

Success codes are number from 1000 upwards so as not to clash with the 8-bit error codes from ESP-IDF.

Tested connection. failure and success on Wemos R32 device, it seems to return as expected.
[wifi_log.txt](https://github.com/micropython/micropython/files/2191225/wifi_log.txt)
Log attached

### CODE REVIEW

**Comment by mitchins** [comment]
@dpgeorge apologies but GitHub created a new PR as the old one was closed.

File: ports/esp32/modnetwork.c [hunk 1]
```diff
@@ -332,12 +337,57 @@ STATIC mp_obj_t esp_disconnect(mp_obj_t self_in) {
...
 
 STATIC mp_obj_t esp_status(size_t n_args, const mp_obj_t *args) {
+    wlan_if_obj_t *self = MP_OBJ_TO_PTR(args[0]);
     if (n_args == 1) {
-        // no arguments: return None until link status is implemented
-        return mp_const_none;
+        if (self->if_id == WIFI_IF_STA) {
+            // Case of no arg is only for the STA interface
+            if (wifi_sta_connected) {
+                return MP_ROM_QSTR(MP_QSTR_STAT_GOT_IP);
```
Comment (dpgeorge) [inline ports/esp32/modnetwork.c]: You want to return the number, like `MP_OBJ_NEW_SMALL_INT(STAT_GOT_IP)`

File: ports/esp32/modnetwork.c [hunk 2]
```diff
@@ -332,12 +337,57 @@ STATIC mp_obj_t esp_disconnect(mp_obj_t self_in) {
...
+            } else if (wifi_sta_connect_requested) {
+                return MP_ROM_QSTR(MP_QSTR_STAT_CONNECTING);
```
Comment (dpgeorge) [inline ports/esp32/modnetwork.c]: Same as above, return `STAT_CONNECTING`.

File: ports/esp32/modnetwork.c [hunk 3]
```diff
@@ -332,12 +337,57 @@ STATIC mp_obj_t esp_disconnect(mp_obj_t self_in) {
...
+            } else {
+                switch (wifi_sta_disconn_reason) {
```
Comment (dpgeorge) [inline ports/esp32/modnetwork.c]: And here you can just return `MP_OBJ_NEW_SMALL_INT(wifi_sta_disconn_reason)`.  Then you don't need any switch statement at all.  That was the reason to reuse the constants from the IDF.

File: ports/esp32/modnetwork.c [hunk 4]
```diff
@@ -55,7 +55,7 @@
 
 NORETURN void _esp_exceptions(esp_err_t e) {
    switch (e) {
-      case ESP_ERR_WIFI_NOT_INIT: 
+      case ESP_ERR_WIFI_NOT_INIT:
```
Comment (mitchins) [inline ports/esp32/modnetwork.c]: Atom keeps stripping the ' ' every time I sae the file.

File: ports/esp32/modnetwork.c [hunk 5]
```diff
@@ -657,6 +687,16 @@ STATIC const mp_rom_map_elem_t mp_module_network_globals_table[] = {
...
+
+    { MP_ROM_QSTR(MP_QSTR_STAT_IDLE), MP_ROM_INT(STAT_IDLE)},
+    { MP_ROM_QSTR(MP_QSTR_STAT_CONNECTING), MP_ROM_INT(STAT_CONNECTING)},
+    { MP_ROM_QSTR(MP_QSTR_STAT_GOT_IP), MP_ROM_INT(STAT_GOT_IP)},
+    //Errors from the ESP-IDF
+    { MP_ROM_QSTR(MP_QSTR_STAT_NO_AP_FOUND), MP_ROM_INT(WIFI_REASON_NO_AP_FOUND)},
+    { MP_ROM_QSTR(MP_QSTR_STAT_WRONG_PASSWORD), MP_ROM_INT(WIFI_REASON_AUTH_FAIL)},
+    { MP_ROM_QSTR(MP_QSTR_STAT_ERROR_BEACON_TIMEOUT), MP_ROM_INT(WIFI_REASON_BEACON_TIMEOUT)},
+    { MP_ROM_QSTR(MP_QSTR_STAT_ERROR_ASSOC_FAIL), MP_ROM_INT(WIFI_REASON_ASSOC_FAIL)},
+    { MP_ROM_QSTR(MP_QSTR_STAT_ERROR_HANDSHAKE_TIMEOUT), MP_ROM_INT(WIFI_REASON_HANDSHAKE_TIMEOUT)},
```
Comment (dpgeorge) [inline ports/esp32/modnetwork.c]: I think for these 3 constants the "ERROR" part is not needed.  Simply `STAT_ASSOC_FAIL` would be enough, etc.

Comment (mitchins) [reply ports/esp32/modnetwork.c]: Agreed, it's a little verbose and it's universally inferred to be an error. Will change

**Comment by dpgeorge** [comment]
Thanks for updating.  I squashed the commits and merged them in 385fa5180663221bbec033c54c37fb38f589579b.  I changed indentation to match the existing code.

**Pull request closed without merging**